velox/type/Filter.cpp

```diff
@@ -458,6 +470,75 @@ bool MultiRange::testBytesRange(
   return false;
 }
 
+std::unique_ptr<Filter> MultiRange::mergeWith(const Filter* other) const {
+  switch (other->kind()) {
+    case FilterKind::kMultiRange: {
+      bool bothNullAllowed = nullAllowed_ && other->testNull();
+      // TODO: verify nanAllowed with the other filter.
+      const MultiRange* multiRangeOther = static_cast<const MultiRange*>(other);
+      std::vector<std::unique_ptr<Filter>> merged;
+      for (auto const& filter : this->filters()) {
+        for (auto const& filterOther : multiRangeOther->filters()) {
+          auto innerMerged = filter->mergeWith(filterOther.get());
+          switch (innerMerged->kind()) {
+            case FilterKind::kAlwaysFalse:
+              break;
+            case FilterKind::kIsNotNull:
+            case FilterKind::kIsNull:
+              // the merged filter is valid when it matches with top-level
+              // nullAllowed.
+              if (innerMerged->testNull() && bothNullAllowed) {
+                merged.push_back(std::move(innerMerged));
+              }
+              break;
+            default:
+              merged.push_back(std::move(innerMerged));
+          }
+        }
+      }
+
+      if (merged.empty()) {
+        return nullOrFalse(bothNullAllowed);
+      } else if (merged.size() == 1) {
+        // If both sides allow null while the merged inner filter does not allow
```

Comment:
We should not consider nullAllowed flags of the inner filters, only top-level nullAllowed flag should be used. Hence, we need to make a clone of merged[0] with nullAllowed = bothNullAllowed